tensornetwork/backends/pytorch/pytorch_backend.py

```diff
@@ -359,6 +353,98 @@ def matmul(self, tensor1: Tensor, tensor2: Tensor) -> Tensor:
+    i.e., the collection of elements of the form a[i, i+offset].
+    If a has more than two dimensions, then the axes specified by
+    axis1 and axis2 are used to determine the 2-D sub-array whose diagonal is
+    returned. The shape of the resulting array can be determined by removing
+    axis1 and axis2 and appending an index to the right equal to the size of the
+    resulting diagonals.
+
+    This function only extracts diagonals. If you
+    wish to create diagonal matrices from vectors, use diagflat.
+
+    Args:
+      tensor: A tensor.
+      offset: Offset of the diagonal from the main diagonal.
+      axis1, axis2: Axis to be used as the first/second axis of the 2D
+                    sub-arrays from which the diagonals should be taken.
+                    Defaults to second-last and last axis (note this
+                    differs from the NumPy defaults).
+    Returns:
+      array_of_diagonals: A dim = min(1, tensor.ndim - 2) tensor storing
+                          the batched diagonals.
+    """
+    if axis1 == axis2:
+      raise ValueError("axis1={axis1} and axis2={axis2} must be different.")
+    return torchlib.diagonal(tensor, offset=offset, dim1=axis1, dim2=axis2)
+
+  def diagflat(self, tensor: Tensor, k: int = 0) -> Tensor:
+    """ Flattens tensor and creates a new matrix of zeros with its elements
+    on the k'th diagonal.
+    Args:
+      tensor: A tensor.
+      k     : The diagonal upon which to place its elements.
+    Returns:
+      tensor: A new tensor with all zeros save the specified diagonal.
+    """
+    return torchlib.diag_embed(tensor, offset=k)
+
+  def trace(self, tensor: Tensor, offset: int = 0, axis1: int = -2,
+            axis2: int = -1) -> Tensor:
+    """Return summed entries along diagonals.
+
+    If tensor is 2-D, the sum is over the
+    diagonal of tensor with the given offset,
+    i.e., the collection of elements of the form a[i, i+offset].
+    If a has more than two dimensions, then the axes specified by
+    axis1 and axis2 are used to determine the 2-D sub-array whose diagonal is
+    summed.
+
+    In the PyTorch backend the trace is always over the main diagonal of the
+    last two entries.
+
+    Args:
+      tensor: A tensor.
+      offset: Offset of the diagonal from the main diagonal.
+              This argument is not supported  by the PyTorch 
+              backend and an error will be raised if they are
+              specified.
+      axis1, axis2: Axis to be used as the first/second axis of the 2D
+                    sub-arrays from which the diagonals should be taken.
+                    Defaults to first/second axis.
+                    These arguments are not supported by the PyTorch
+                    backend and an error will be raised if they are
+                    specified.
+    Returns:
+      array_of_diagonals: The batched summed diagonals.
+    """
+    if offset != 0:
```

Comment:
We could actually also have an offset by using contraction with an offset identity matrix.

Reply:
Let's leave it for later though